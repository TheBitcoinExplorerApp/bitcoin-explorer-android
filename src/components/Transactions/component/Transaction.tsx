/* eslint-disable import/no-extraneous-dependencies */
import { useState } from "react";
import Modal from "src/components/Modal/Modal";
import ModalServices from "src/services/ModalServices";
import useAppDataStore from "src/context/DataProvider";
import { TransactionType } from "src/components/Modal/types";
import { View, Text, StyleSheet, Pressable } from "react-native";
import { TransactionInfo } from "../types";

type transaction = {
  transactionId: string;
  value: number;
  fee: number;
};

type TransactionProps = {
  transactions: transaction[];
};

const styles = StyleSheet.create({
  transactionsContainer: {
    gap: 16,
  },
  container: {
    flexDirection: "row",
    justifyContent: "space-evenly",
    backgroundColor: "#1d2133",
    borderRadius: 7,
    paddingVertical: 10,
  },
  transactionInfoContainer: {
    alignItems: "center",
    gap: 8,
  },
  transactionId: {
    color: "#DF7800",
    fontWeight: "500",
  },
  transactionInfoLabel: {
    color: "white",
    fontSize: 13,
    fontWeight: "500",
  },
  transactionInfoValue: {
    color: "white",
    fontWeight: "500",
  },
});

export default function Transaction(props: TransactionProps) {
  const { transactions } = props;
  const { i18n } = useAppDataStore();
  const [modalVisible, setModalVisible] = useState(false);
  const [transactionInfo, setTransactionInfo] = useState<TransactionInfo>({
    transactionData: {} as TransactionType,
    transactionHash: "",
  });

  const getTransactionInfo = async (transactionHash: string) => {
    const result =
      await ModalServices.getTransactionTransactionsInfo(transactionHash);

    setTransactionInfo({
      transactionHash: result.transactionId,
      transactionData: result,
    });
  };

  return (
    <>
      <View style={styles.transactionsContainer}>
        {transactions.map((transaction) => {
          const { transactionId, fee, value } = transaction;

          return (
            <Pressable
              onPress={() => {
                setTransactionInfo({
                  ...transactionInfo,
                  transactionHash: transactionId,
                });
                getTransactionInfo(transactionId);
                setModalVisible(true);
              }}
              key={transactionId}
            >
              <View style={styles.container}>
                <View style={styles.transactionInfoContainer}>
                  <Text style={styles.transactionInfoLabel}>
                    {i18n.t("transaction_id")}
                  </Text>
                  <Text style={styles.transactionId}>
                    {transactionId.slice(0, 14).concat("...")}
                  </Text>
                </View>

                <View style={styles.transactionInfoContainer}>
                  <Text style={styles.transactionInfoLabel}>
                    {i18n.t("amount")}
                  </Text>
                  <Text style={styles.transactionInfoValue}>
                    {value / 100000000} BTC
                  </Text>
                </View>
                <View style={styles.transactionInfoContainer}>
                  <Text style={styles.transactionInfoLabel}>
                    {i18n.t("fee")}
                  </Text>
                  <Text style={styles.transactionInfoValue}>
                    {fee.toLocaleString()} sats
                  </Text>
                </View>
              </View>
            </Pressable>
          );
        })}
      </View>

      <Modal
        isVisible={modalVisible}
        modalType="Transaction"
        transactionHash={transactionInfo.transactionHash}
        transactionInfo={transactionInfo.transactionData}
        handleModalClose={() => {
          setModalVisible(false);
          setTransactionInfo({
            ...transactionInfo,
            transactionData: {} as TransactionType,
            transactionHash: "",
          });
        }}
      />
    </>
  );
}
