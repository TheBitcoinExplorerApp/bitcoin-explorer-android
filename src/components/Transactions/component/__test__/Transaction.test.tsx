// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from "react-test-renderer";
import Transaction from "../Transaction";

describe("Transaction", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Transaction
          transactions={[{ fee: 10, transactionId: "#ABC12", value: 10 }]}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
