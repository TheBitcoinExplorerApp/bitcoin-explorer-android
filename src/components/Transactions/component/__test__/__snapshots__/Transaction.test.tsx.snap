// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction renders correctly 1`] = `
<View
  style={
    {
      "gap": 16,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      style={
        {
          "backgroundColor": "#1d2133",
          "borderRadius": 7,
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "paddingVertical": 10,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "gap": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 13,
              "fontWeight": "500",
            }
          }
        >
          ID Transação
        </Text>
        <Text
          style={
            {
              "color": "#DF7800",
              "fontWeight": "500",
            }
          }
        >
          #ABC12...
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "gap": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 13,
              "fontWeight": "500",
            }
          }
        >
          Valor
        </Text>
        <Text
          style={
            {
              "color": "white",
              "fontWeight": "500",
            }
          }
        >
          1e-7
           BTC
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "gap": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 13,
              "fontWeight": "500",
            }
          }
        >
          Taxa
        </Text>
        <Text
          style={
            {
              "color": "white",
              "fontWeight": "500",
            }
          }
        >
          10
           sats
        </Text>
      </View>
    </View>
  </View>
</View>
`;
